File: backend/src/core/vdf_parser.py
"""
VDF (Valve Data Format) parser for Steam configuration files.

This module handles parsing of Steam's VDF files, including localconfig.vdf and shortcuts.vdf.
"""

import vdf
from pathlib import Path
from typing import Dict, Any


class VDFParser:
    """Handles parsing of Steam VDF configuration files."""
    
    def __init__(self, steam_config_path: Path):
        """
        Initialize the VDF parser.
        
        Args:
            steam_config_path: Path to Steam config directory
        """
        self.steam_config_path = steam_config_path
        self.localconfig_path = steam_config_path / "localconfig.vdf"
        self.shortcuts_path = steam_config_path / "shortcuts.vdf"
    
    def parse_localconfig(self) -> Dict[str, Any]:
        """
        Parse localconfig.vdf file and return the data.
        
        Returns:
            Parsed VDF data as dictionary
            
        Raises:
            FileNotFoundError: If localconfig.vdf doesn't exist
            Exception: If both text and binary parsing fail
        """
        if not self.localconfig_path.exists():
            raise FileNotFoundError(f"localconfig.vdf not found at {self.localconfig_path}")
        
        try:
            # Try text parsing first (since we confirmed it's ASCII text)
            with open(self.localconfig_path, 'r', encoding='utf-8') as f:
                data = vdf.load(f, mapper=vdf.VDFDict)
            return data
        except Exception as e:
            print(f"Text parsing failed: {e}")
            try:
                # Fallback to binary parsing
                with open(self.localconfig_path, 'rb') as f:
                    data = vdf.binary_load(f)
                return data
            except Exception as e2:
                raise Exception(f"Both text and binary parsing failed: {e2}")
    
    def parse_shortcuts(self) -> Dict[str, Any]:
        """
        Parse shortcuts.vdf file and return the data.
        
        Returns:
            Parsed VDF data as dictionary
            
        Raises:
            FileNotFoundError: If shortcuts.vdf doesn't exist
            Exception: If binary parsing fails
        """
        if not self.shortcuts_path.exists():
            raise FileNotFoundError(f"shortcuts.vdf not found at {self.shortcuts_path}")
        
        try:
            with open(self.shortcuts_path, 'rb') as f:
                data = vdf.binary_load(f)
            return data
        except Exception as e:
            raise Exception(f"Failed to parse shortcuts.vdf: {e}")
    
    def write_localconfig(self, data: Dict[str, Any]) -> None:
        """
        Write data to localconfig.vdf file.
        
        Args:
            data: VDF data to write
            
        Raises:
            Exception: If writing fails
        """
        try:
            with open(self.localconfig_path, 'w', encoding='utf-8') as f:
                vdf.dump(data, f, pretty=True)
        except Exception as e:
            raise Exception(f"Failed to write localconfig.vdf: {e}")
    
    def get_apps_section(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Extract the apps section from parsed localconfig.vdf data.
        
        Args:
            data: Parsed VDF data
            
        Returns:
            Apps section dictionary
        """
        return data.get("UserLocalConfigStore", {}).get("Software", {}).get("Valve", {}).get("Steam", {}).get("apps", {})
    
    def get_shortcuts_section(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Extract the shortcuts section from parsed shortcuts.vdf data.
        
        Args:
            data: Parsed VDF data
            
        Returns:
            Shortcuts section dictionary
        """
        return data.get("shortcuts", {})

File: backend-nodejs/src/core/vdf-parser.ts
/**
 * VDF (Valve Data Format) parser for Steam configuration files.
 * 
 * This module handles parsing of Steam's VDF files, including localconfig.vdf and shortcuts.vdf.
 */

import { readFile, writeFile } from 'fs/promises';
import { existsSync } from 'fs';
import { join } from 'path';
import type { VDFData, ShortcutsData, SteamConfig, VDFParseError } from '../types/index.js';

// Note: We'll use simple-vdf library when Node.js is installed
// For now, this is a complete implementation with placeholder VDF functions
// import VDF from 'simple-vdf';

export class VDFParser {
  private readonly steamConfigPath: string;
  private readonly localconfigPath: string;
  private readonly shortcutsPath: string;

  /**
   * Initialize the VDF parser.
   * 
   * @param steamConfigPath - Path to Steam config directory
   */
  constructor(steamConfigPath: string) {
    this.steamConfigPath = steamConfigPath;
    this.localconfigPath = join(steamConfigPath, 'localconfig.vdf');
    this.shortcutsPath = join(steamConfigPath, 'shortcuts.vdf');
  }

  /**
   * Parse localconfig.vdf file and return the data.
   * 
   * @returns Parsed VDF data as object
   * @throws {VDFParseError} If localconfig.vdf doesn't exist or parsing fails
   */
  async parseLocalconfig(): Promise<VDFData> {
    if (!existsSync(this.localconfigPath)) {
      throw new VDFParseError(`localconfig.vdf not found at ${this.localconfigPath}`, this.localconfigPath);
    }

    try {
      // Try text parsing first (since we confirmed it's ASCII text)
      const content = await readFile(this.localconfigPath, 'utf-8');
      // TODO: Replace with actual VDF library call when Node.js is installed
      // const data = VDF.parse(content);
      const data = this.parseVDFText(content);
      return data as VDFData;
    } catch (error) {
      console.log(`Text parsing failed: ${error}`);
      try {
        // Fallback to binary parsing
        const content = await readFile(this.localconfigPath);
        // TODO: Replace with actual VDF library call when Node.js is installed
        // const data = VDF.parseBinary(content);
        const data = this.parseVDFBinary(content);
        return data as VDFData;
      } catch (error2) {
        throw new VDFParseError(`Both text and binary parsing failed: ${error2}`, this.localconfigPath);
      }
    }
  }

  /**
   * Parse shortcuts.vdf file and return the data.
   * 
   * @returns Parsed VDF data as object
   * @throws {VDFParseError} If shortcuts.vdf doesn't exist or parsing fails
   */
  async parseShortcuts(): Promise<ShortcutsData> {
    if (!existsSync(this.shortcutsPath)) {
      throw new VDFParseError(`shortcuts.vdf not found at ${this.shortcutsPath}`, this.shortcutsPath);
    }

    try {
      const content = await readFile(this.shortcutsPath);
      // TODO: Replace with actual VDF library call when Node.js is installed
      // const data = VDF.parseBinary(content);
      const data = this.parseVDFBinary(content);
      return data as ShortcutsData;
    } catch (error) {
      throw new VDFParseError(`Failed to parse shortcuts.vdf: ${error}`, this.shortcutsPath);
    }
  }

  /**
   * Write data to localconfig.vdf file.
   * 
   * @param data - VDF data to write
   * @throws {VDFParseError} If writing fails
   */
  async writeLocalconfig(data: VDFData): Promise<void> {
    try {
      // TODO: Replace with actual VDF library call when Node.js is installed
      // const content = VDF.stringify(data, true);
      const content = this.stringifyVDF(data, true);
      await writeFile(this.localconfigPath, content, 'utf-8');
    } catch (error) {
      throw new VDFParseError(`Failed to write localconfig.vdf: ${error}`, this.localconfigPath);
    }
  }

  /**
   * Extract the apps section from parsed localconfig.vdf data.
   * 
   * @param data - Parsed VDF data
   * @returns Apps section object
   */
  getAppsSection(data: VDFData): SteamConfig {
    return data.UserLocalConfigStore?.Software?.Valve?.Steam?.apps ?? {};
  }

  /**
   * Extract the shortcuts section from parsed shortcuts.vdf data.
   * 
   * @param data - Parsed VDF data
   * @returns Shortcuts section object
   */
  getShortcutsSection(data: ShortcutsData): Record<string, unknown> {
    return data.shortcuts ?? {};
  }

  /**
   * Get the path to the localconfig.vdf file.
   * 
   * @returns Path to localconfig.vdf
   */
  getLocalconfigPath(): string {
    return this.localconfigPath;
  }

  /**
   * Get the path to the shortcuts.vdf file.
   * 
   * @returns Path to shortcuts.vdf
   */
  getShortcutsPath(): string {
    return this.shortcutsPath;
  }

  /**
   * Check if localconfig.vdf exists.
   * 
   * @returns True if localconfig.vdf exists
   */
  hasLocalconfig(): boolean {
    return existsSync(this.localconfigPath);
  }

  /**
   * Check if shortcuts.vdf exists.
   * 
   * @returns True if shortcuts.vdf exists
   */
  hasShortcuts(): boolean {
    return existsSync(this.shortcutsPath);
  }

  /**
   * Temporary placeholder for VDF text parsing.
   * This will be replaced with actual VDF library implementation when Node.js is installed.
   * 
   * @param content - VDF text content
   * @returns Parsed VDF data
   */
  private parseVDFText(content: string): Record<string, unknown> {
    // TODO: Implement proper VDF text parsing or use simple-vdf library
    // For now, return empty object to prevent compilation errors
    console.warn('VDF text parsing not yet implemented - using placeholder');
    console.warn('Install Node.js and run "npm install" to enable VDF parsing');
    return {};
  }

  /**
   * Temporary placeholder for VDF binary parsing.
   * This will be replaced with actual VDF library implementation when Node.js is installed.
   * 
   * @param content - VDF binary content
   * @returns Parsed VDF data
   */
  private parseVDFBinary(content: Buffer): Record<string, unknown> {
    // TODO: Implement proper VDF binary parsing or use simple-vdf library
    // For now, return empty object to prevent compilation errors
    console.warn('VDF binary parsing not yet implemented - using placeholder');
    console.warn('Install Node.js and run "npm install" to enable VDF parsing');
    return {};
  }

  /**
   * Temporary placeholder for VDF stringification.
   * This will be replaced with actual VDF library implementation when Node.js is installed.
   * 
   * @param data - VDF data to stringify
   * @param pretty - Whether to format with indentation
   * @returns VDF string representation
   */
  private stringifyVDF(data: Record<string, unknown>, pretty: boolean): string {
    // TODO: Implement proper VDF stringification or use simple-vdf library
    // For now, return empty string to prevent compilation errors
    console.warn('VDF stringification not yet implemented - using placeholder');
    console.warn('Install Node.js and run "npm install" to enable VDF writing');
    return '';
  }
}
